src/js/libcs/Render2D.js

```diff
@@ -500,59 +500,49 @@ Render2D.drawpoint = function(pt) {
 };
 
 Render2D.drawline = function(homog) {
-    var na = CSNumber.abs(homog.value[0]).value.real;
-    var nb = CSNumber.abs(homog.value[1]).value.real;
-    var nc = CSNumber.abs(homog.value[2]).value.real;
-    var divi;
+    if (!List._helper.isAlmostReal(homog))
+        return;
 
-    if (na >= nb && na >= nc) {
-        divi = homog.value[0];
-    }
-    if (nb >= na && nb >= nc) {
-        divi = homog.value[1];
-    }
-    if (nc >= nb && nc >= na) {
-        divi = homog.value[2];
-    }
-    var a = CSNumber.div(homog.value[0], divi);
-    var b = CSNumber.div(homog.value[1], divi);
-    var c = CSNumber.div(homog.value[2], divi); //TODO Realitycheck einbauen
-
-    var l = [
-        a.value.real,
-        b.value.real,
-        c.value.real
-    ];
-    var b1, b2;
-    if (Math.abs(l[0]) < Math.abs(l[1])) {
-        b1 = [1, 0, 30];
-        b2 = [-1, 0, 30];
-    } else {
-        b1 = [0, 1, 30];
-        b2 = [0, -1, 30];
-    }
-    var erg1 = [
-        l[1] * b1[2] - l[2] * b1[1],
-        l[2] * b1[0] - l[0] * b1[2],
-        l[0] * b1[1] - l[1] * b1[0]
-    ];
-    var erg2 = [
-        l[1] * b2[2] - l[2] * b2[1],
-        l[2] * b2[0] - l[0] * b2[2],
-        l[0] * b2[1] - l[1] * b2[0]
-    ];
-
-    var pt1 = {
-        x: erg1[0] / erg1[2],
-        y: erg1[1] / erg1[2]
-    };
-    var pt2 = {
-        x: erg2[0] / erg2[2],
-        y: erg2[1] / erg2[2]
+    var n = List.normalizeMax(homog);
+    var a = n.value[0].value.real;
+    var b = n.value[1].value.real;
+    var c = n.value[2].value.real;
 
+    var distNeg = function(x, y) {
+        return x * a + y * b + c < 0;
     };
 
-    Render2D.drawsegcore(pt1, pt2);
+    // do not draw beyond canvas boundary
+    var m = csport.drawingstate.matrix;
+    var xMin = (m.b * m.ty - m.d * m.tx) / m.det;
+    var xMax = (m.b * (m.ty - csh) - m.d * (m.tx - csw)) / m.det;
+    var yMax = (m.c * m.tx - m.a * m.ty) / m.det;
+    var yMin = (m.c * (m.tx - csw) - m.a * (m.ty + csh)) / m.det;
```

Comment:
> But in the end we need to transform the arguments of `Render2D.drawsegcore` back to user space.

I see no reason why the arguments to `drawsegcore` *have* to be user space coordinates. We could make them canvas space, and let the caller of `drawsegcore` handle conversion. Or move the bulk of `drawsegcore` to some new function, and let `drawsegcore` do the conversion and then call the new function.

> On the other hand, we should of course also clip segments. So we will have some duplicate code in any case, as far as I see it.

When doing segment clipping, we have to be particularly aware of the fact that a segment between two (almost) identical points should be supported, even though the line connecting those is not uniquely defined. So I'd probably build segment clipping on the notion of whether the endpoints of a segment are separated by a given canvas border. So that you don't compute points of intersection unneccessarily. I'd say this probably should go into some function that calls `drawsegcore`, not into `drawsegcore` itself as it is now. If you make those two functions mentioned above, perhaps the outer one of them is the right place for segment clipping?

Not sure what to do about rays, though. Do we need a third code path, or does it make sense to somehow treat them as a special case of lines or segments? We will encounter rays e.g. in the case of segments that pass through infinity.